File: demo/src/main/java/com/moka/TestViewModel.java
package com.moka;

import android.content.Context;
import android.os.Handler;
import android.view.View;
import android.widget.Toast;

import com.moka.annotations.BindTo;
import com.moka.annotations.BindsTo;
import com.moka.mvvm.Observable;
import com.moka.mvvm.ViewModel;

import org.jetbrains.annotations.NotNull;

/**
 * Created by wangjinpeng on 2017/11/9.
 */

public class TestViewModel extends ViewModel {

    private Handler handler = new Handler();

    @BindTo(TestViewProtocol.helloWordText)
    private Observable<String> info = new Observable<>("hello world !!!!!!");

    @BindsTo({
            @BindTo(value = TestViewProtocol.buttonEnable, field = "enable"),
            @BindTo(value = TestViewProtocol.buttonText, field = "text")
    })
    private Observable<ButtonInfo> buttonInfo = new Observable<>();

    public TestViewModel(@NotNull Context context) {
        super(context);
    }

    @BindTo(TestViewProtocol.onButton2Click)
    public void onButton2Click(View view) {
        Toast.makeText(getContext(), "点击了button2", Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onCreate() {
        super.onCreate();
        handler.postDelayed(new Runnable() {
            @Override
            public void run() {
                info.set("和理论绿绿绿绿绿绿绿绿绿绿");
                buttonInfo.set(new ButtonInfo("延时设置的button文案", false));
            }
        }, 2000);
    }

    public Observable<String> getInfo() {
        return info;
    }

    public Observable<ButtonInfo> getButtonInfo() {
        return buttonInfo;
    }
}

File: demo/src/main/java/com/moka/TestViewModel.kt
package com.moka

import android.content.Context
import android.os.Handler
import android.view.View
import android.widget.Toast

import com.moka.annotations.ObserveBy
import com.moka.mvvm.MVVM
import com.moka.mvvm.Observable
import com.moka.mvvm.ViewModel

/**
 * Created by wangjinpeng on 2017/11/9.
 */

class TestViewModel(context: Context) : ViewModel(context), TestViewProtocol {

    private val handler = Handler()

    @ObserveBy(TestViewProtocol.getHelloWordText)
    val info = Observable("hello world !!!!!!")

    @ObserveBy(TestViewProtocol.getButtonEnable, TestViewProtocol.getButtonText)
    val buttonInfo = Observable<ButtonInfo>()

    override fun onCreate() {
        super.onCreate()
        handler.postDelayed({
            info.set("和理论绿绿绿绿绿绿绿绿绿绿")
            buttonInfo.set(ButtonInfo("延时设置的button文案", false))
        }, 2000)
    }

    override fun onButton2Click(view: View?) {
        Toast.makeText(context, "点击了button2", Toast.LENGTH_SHORT).show()
    }

    fun getHelloWortText(view: View): Observable<String>? {
        return null
    }

    override fun getHelloWordText(): Observable<String> {
        return info
    }

    override fun getButtonText(): Observable<String> {
        val buttonInfo = this.buttonInfo.get()
        return MVVM
                .wrap<String>(this, buttonInfo?.text)
                .describe(ButtonInfo::class.java, "text")
                .create()
    }

    override fun getButtonEnable(): Observable<Boolean> {
        val buttonInfo = this.buttonInfo.get()
        return MVVM.wrap<Boolean>(this, buttonInfo?.isEnable)
                .describe(ButtonInfo::class.java, "enable")
                .create()
    }

    override fun requestData(boolean: Boolean) {

    }
}
